compiler/rustc_parse/src/lexer/mod.rs

```diff
@@ -732,6 +734,102 @@ impl<'psess, 'src> StringReader<'psess, 'src> {
+        self.pos = start + BytePos(1);
+        self.cursor = Cursor::new(&str_before[1..]);
+        token::Pound
+    }
+
+    /// Detect guarded string literal syntax
+    ///
+    /// RFC 3598 reserved this syntax for future use. As of Rust 2024,
+    /// using this syntax produces an error. In earlier editions, however, it
+    /// only results in an (allowed by default) lint, and is treated as
+    /// separate tokens.
+    fn report_guarded_str(&mut self, start: BytePos, str_before: &'src str) -> TokenKind {
+        let span = self.mk_sp(start, self.pos);
+        let edition2024 = span.edition().at_least_rust_2024();
+
+        // Prior to edition 2024, check the previous char, so we don't
+        // lint for each `#` before and after the string.
+        if !edition2024
+            && start > self.start_pos
+            && matches!(self.src.as_bytes()[self.src_index(start) - 1], b'#' | b'"')
+        {
+            return self.split_guarded_str_maybe(start, str_before);
+        }
+
+        let mut maybe_cursor = Cursor::new(str_before);
+
+        let (span, space_span, unterminated) =
+            if let Some(rustc_lexer::GuardedStr { n_hashes, terminated, token_len }) =
+                maybe_cursor.guarded_double_quoted_string()
+            {
+                let end = start + BytePos(token_len);
+                let span = self.mk_sp(start, end);
+                let str_start = start + BytePos(n_hashes);
+                let space_span = self.mk_sp(str_start, str_start);
+
+                if edition2024 {
+                    self.cursor = maybe_cursor;
+                    self.pos = end;
+                }
+
+                let unterminated = if edition2024 && !terminated { Some(str_start) } else { None };
+
+                (span, space_span, unterminated)
+            } else {
+                // We should only get here in the `##+` case.
+                debug_assert_eq!(self.str_from_to(start, start + BytePos(2)), "##");
+
+                let space_pos = start + BytePos(1);
+                let space_span = self.mk_sp(space_pos, space_pos);
+
+                (span, space_span, None)
+            };
+        if edition2024 {
+            let expn_data = span.ctxt().outer_expn_data();
+
+            let sugg = if expn_data.is_root() {
+                Some(errors::GuardedStringSugg(space_span))
+            } else {
+                None
+            };
+
+            // In Edition 2024 and later, emit a hard error.
+            let err = self.dcx().emit_err(errors::ReservedString { span, sugg });
+
+            if let Some(str_start) = unterminated {
+                // Only a fatal error if string is unterminated.
+                self.dcx()
+                    .struct_span_fatal(
+                        self.mk_sp(str_start, self.pos),
+                        "unterminated double quote string",
+                    )
+                    .with_code(E0765)
+                    .emit()
```

Comment:
There is a double error here, one for reservation and another for unterminated string, one would be enough (with the fatal one as a priority).